### ISSUE TITLE
[Feature]: Step slider for `wandb.Table` output

### ISSUE THREAD

**Issue by johnPertoft** [wandb/wandb#6286]
### Description

I'm logging a `wandb.Table` after every training+evaluation epoch. The rows of this table are sampled rows from an evaluation set that I'd like to I'd like to visualize predictions for at different epochs. I would like to have a slider to see the output at each epoch step.

All my tables are available under the artifacts tab but it would be nice to have a slider to go between them in the same page as the rest of the logged metrics.

If there already exists a way to get *something like this* I'd be interested in this as well.

### Suggested Solution

A slider to quickly go between different iterations of a `wandb.Table` under that same key.

**Comment by johnPertoft** [comment]
Or is the recommended way to add rows from subsequent epochs to the same table and then do a groupby on input for example?

**Comment by sydholl** [comment]
WandB Internal User commented: 
johnPertoft commented: 
Or is the recommended way to add rows from subsequent epochs to the same table and then do a groupby on input for example?

**Comment by umakrishnaswamy** [comment]
Hey @johnPertoft, currently we do not support step sliders for wandb Tables, but I'd be happy to make this feature request for you. 

I really like your idea of adding rows from subsequent epochs and grouping by values, and this is what I would recommend as well. Another idea could be logging these values as separate tables/artifacts and viewing them in reports (would be a bit more cumbersome to go through, so I'd still recommend the former more so). Please let me know if you have other questions!

**Comment by kxhit** [comment]
Hi, @umakrishnaswamy  I also want the feature of supporting step slider for table of images. As the list of images are limited by 108 images, I change my wandb code to use table of images, but found out there is no step slider so I can see the changes of my images across training steps. I really want to have this feature and please let me know. Thanks!

**Comment by umakrishnaswamy** [comment]
@kxhit - Noted! Will be sure to update this thread if any progress arises on the feature request

**Comment by Xiashangning** [comment]
+1, I also encountered this problem. When training generative model, I would like to keep en eye on the generative ability of the model along the training process. However, as the total number of images is larger than 108, I switched my code to use table of images. However, in this case I can only see the last updated images and can not compare them with previous steps. This is truly upsetting... This feature is reasonable even for the regular tables I think.
PS: If currently the step slider is not available, is there any way to filter the table's output by `_step` in the query ? For example support query like this `runs.history["_step"==500]["images"]`

**Comment by oliver-batchelor** [comment]
Same.

**Comment by nickojelly** [comment]
+1, would be very useful when plotting scatterplots of wandb tables

**Comment by OmerRochman** [comment]
+1, it would be great also for autoregressive models where you unroll the model for some timesteps at a given training iteration and want to measure and plot different metrics computed on the trajectory as a function of the timestep and see how those change over training.

**Comment by 01tot10** [comment]
Ran into this issue earlier this week, so definitely +1.

I'm a bit confused, since the [documentation](https://docs.wandb.ai/guides/track/log/log-tables#log-a-table-to-a-run) states that:
> Each time a table is logged to the same key, a new version of the table is created and stored in the backend. This means you can log the same table across multiple training steps to see how model predictions improve over time, or compare tables across different runs, as long as they're logged to the same key.

I was assuming the ability to "_see how model predictions improve over time_" would literally mean that there's the slider for dynamically adjusting the table as a function of training step, but that seems to not be the case. What sort of a workflow would this sentence in the documentation be referring to?

**Comment by umakrishnaswamy** [comment]
hey @01tot10,

if you are logging text to a table, you can use our Text table feature, which would include a slider to step through the eval steps and view the table at those steps (caveat: this doesn’t work for all data types) 

```import wandb

# Initialize a wandb run
wandb.init(project="text-table", entity="uma-wandb")

# Define the number of steps for demonstration purposes
num_steps = 5  # Replace with the actual number of steps in your process

for step in range(num_steps):
    # Create a wandb.Table() object with the specified dimensions
    table = wandb.Table(columns=["Column 1", "Column 2"])
    
    # Populate the table with 10 rows and some text
    for row in range(10):
        table.add_data(f"Text {row} step {step} col 1", f"Text {row} step {step} col 2")
    
    # Log the table at the current step
    wandb.log({f"table": table}, step=step)```

example usage if you’d like to try this out - this would show up as a single table in the wandb UI with a step slider
link to [workspace](https://wandb.ai/uma-wandb/text-table/runs/1nkulazj?nw=nwuserumawandb) to check it out

**Comment by 01tot10** [comment]
Hey @umakrishnaswamy! 

Thanks for getting back on the topic.

Judging by your answer, it sounds the workflow exists for "tables of text", but not for all data types.
I was expecting to see a slider in the workspace that was linked in the earlier answer, but it's not obvious to me, where it should appear. See this screenshot:

![Screenshot 2024-05-17 at 9 10 16](https://github.com/wandb/wandb/assets/35490779/347df600-f6cb-473a-9577-cd3cc9b07c75)

I see the two columns and 10 words worth of dummy text, but only for the latest iteration. If I'm missing something obvious in the UI, do let me know!

Anyways, I've so far tried to log figures and audio examples, which are typical in the problems I'm usually working, and it would be great to have the slider feature implemented for those data types too!

**Comment by MBakirWB** [comment]
@01tot10 Currently when you log to the same wandb table key (name) across multiple runs, wandb concatenates all data across all runs at the project level and we continue to append this data. This is the reason why you are only seeing one table available for selection. There are two approach to how you can visualize this data. One is to render the table as a `Combined Table` off all rows of data and the other is `Paginated Tables` where the table is versioned/sectioned by run. Tables don't currently have sliders. 

See the video below of a table that's logged 30 rows of data across three runs. I can control how many rows to display by changing run visibility and also by changing the rendering to section of the table.. Hope this clarifies the behavior on your end. 


<img src="https://github.com/wandb/wandb/assets/104453360/edafc7bb-8e37-479f-84c3-3fb5af0fce4b" alt="concat_rows" width="400" height="300">

<img src="https://github.com/wandb/wandb/assets/104453360/bf1764b5-83b4-4b38-ae65-eea27080a3aa" alt="paginated_table" width="400" height="300" >

https://github.com/wandb/wandb/assets/104453360/5581846b-c5f8-4475-ad17-f6c1f59ded00

**Comment by jin-zhe** [comment]
Late to the discussion. Tabular presentations are extremely useful for organizing various information so things are clear at-a-glance. Consolidating tabular data under one panel and having a step-slider to scrub through these tables is therefore an extremely important and necessary feature.

It's a great pity that wandb doesn't support a simple table API without all the bells and whistles of `wandb.Table`. Currently my data (which I would ideally consolidate under one table) are distributed across different panels and stepping through means I'll have to scrub the slider across every panel. If there's a better way, please let me know!

I'm currently using a simple hack to log the table as a html instead since that comes with the step slider. The function currently supports text and images with CSS styling controls. If it might be of use to anyone, here's the gist: https://gist.github.com/jin-zhe/8f8b9185f61207ab2a3c78aa096b3588

Hope there's an official API for a simple table with step slider soon!

**Comment by ArtsiomWB** [comment]
Hi @jin-zhe! Thank you for the follow-up and a great workaround. Looks like this feature request got buried. I'll resubmit another feature request regarding the slider tables to our engineering team. 

For now, the combined table approach, as Mohammad described above, is the recommended workflow. However, I understand this may not be ideal if you're looking to compare tables one-to-one across different steps. I'll make sure to share this feedback with our engineering team as well.

**Comment by carlthome** [comment]
Also just ran into this confusion in PyTorch Lightning where I assumed that using `self.logger.log_table` with the same key within my `LightningModule` would default to having a step slider in the W&B UI.

**Comment by lqrhy3** [comment]
+1 here

**Comment by thomasgorman-degas** [comment]
+1 here

**Comment by PhilippDahlinger** [comment]
One workaround which I use is to create a Plotly Figure out of your data. If you want to plot a line plot per step, you can do for example the following:

```
fig = go.Figure()
fig.add_trace(go.Scatter(x=steps, y=values, mode='lines', name='Values'))
fig.update_layout(title="title,
                  xaxis_title="Steps",
                  yaxis_title="yaxis desc",
                  template="plotly_white")
wandb.log(fig)
```
I assume you can also log images that way. However, the loading of the figure is sometimes a bit slow. I noticed that it can slow down if you use it extensively. (However, I plotted whole animations like that using multiple steps, this was likely the reason).

Hope that helps!

**Comment by zhaohm14** [comment]
+1. Need this feature

**Comment by mickelliu** [comment]
+1, this feature would be very help to LLM research

**Comment by ArtsiomWB** [comment]
Thank you for the +1s guys, I apologize for the no updates. An engineer on our tables team is currently working on the step slider for the table!

**Comment by ShayanTalaei** [comment]
+1

**Comment by Srinivas-R** [comment]
+1

**Comment by edbeeching** [comment]
Hi @ArtsiomWB , do you have any updates on this feature? The team working on the [open-r1 reproduction](https://github.com/huggingface/open-r1) would really appreciate this.

**Comment by ArtsiomWB** [comment]
Hey guys! quick update! We finally have a step slider in W&B tables! You can access it using the `Runs History Table Stepper.` 

These tables let you view your logged tables across multiple steps!
Here is a very quick video since github is not letting me record anything longer:

https://github.com/user-attachments/assets/e6b53e05-cdaa-4290-aef8-eb7c883f4288

For folks who can't watch it, after logging a table across different steps, you can now go into your run, and set the table to the `runs` query. 

<img width="532" alt="Image" src="https://github.com/user-attachments/assets/03d561c2-d086-4dc3-949d-c72b39b93db1" />

From here you will need to click into the `Query Panel Settings` and select the `Run History Tables Stepper`

<img width="409" alt="Image" src="https://github.com/user-attachments/assets/b87b257c-0854-4232-a93b-8b9ecaf2d779" />

Then, you'll need to select the table name you are interested in displaying

<img width="323" alt="Image" src="https://github.com/user-attachments/assets/df15d40d-ff28-4cda-8c6f-3219331e8589" />

After selecting it, you should see the step slider for your table!

<img width="478" alt="Image" src="https://github.com/user-attachments/assets/8bb0379f-739e-4526-a391-bab8b152e408" />

**Comment by mickelliu** [comment]
Hi @ArtsiomWB, I have tried it out and it worked, thank you!
I think this is a very nice step forward but I do have additional concerns:
- Now that I log each table individually I lost the ability to stack them into one giant table :( and I can't find a query panel that is able to do that.
- I don't believe that the slider can skip empty steps atm?

**Comment by ArtsiomWB** [comment]
Thank you for the feedback!

> Now that I log each table individually I lost the ability to stack them into one giant table :( and I can't find a query panel that is able to do that.

Could you please give me an example of what you are referring to here? I want to make sure that I know exactly what your desired workflow is, incase I'll need to send it directly to eng :)

> I don't believe that the slider can skip empty steps atm?

That is correct, and I can make another feature request regarding this.

**Comment by mickelliu** [comment]
@ArtsiomWB I think in order for the slider to work, you need to `run.log` a new table every step. However when you run `run.summary['table']`, you will just see the last iteration of the table. But since we have all the tables uploaded to the wandb already, why can't you just join them together on the web interface? I really want to have both the slider and joined table across all iterations of the tables.

**Comment by karthi0804** [comment]
Currently, there is no way to get the slider when I view the table as a plot . Any suggestions? Sliding through the plots.

**Comment by ArtsiomWB** [comment]
Oh I see, @mickelliu! You are interested in combining multiple tables into one that you log over multiple steps and then use the step slider to go through the combined table at once, correct? 

> Currently, there is no way to get the slider when I view the table as a plot . Any suggestions? I want to slide through the epochs to view a 2d bar chart like media.

I dont think this is possible at the moment, @karthi0804. I'll go ahead and submit both of these feature requests to our eng team.

**Comment by mickelliu** [comment]
@ArtsiomWB Correct, really wish to have both options without extra steps. 
(1) joined all tables by rows so I can later do some filtering to see how the same prompt-generation pairs evolve over training iterations, and 
(2) at each time step, have a view on the current batch, and jump between the iterations using the slider.

Before the slider feature is implmenented by you guys, I have been using the `run.summary['table']` panel to visualize my result, but it requires keeping a joined table locally, prepend the iteration number, and then upload the table at each time step. However it uploads repeated data which unnecessarily consumes a lot of space.

**Comment by ArtsiomWB** [comment]
Thank you for the feedback and feature requests. I have gone ahead and sent them over to our engineering team. 

I'll let you guys know as soon as I get any updates regarding this on my side. 

Regarding any further questions about the step slider tables and feature requests, please go ahead and create a new thread/ticket so it is easier to track and we can dig into it together.

**Comment by felipemello1** [comment]
thanks for the progress here! Just sharing my use case, which is similar:

1. For every training step, i log the prompt/answer/rewards of the model.
2. I want to be able to visualize/export it as a single concatenated table

Currently i see this:

<img width="1105" alt="Image" src="https://github.com/user-attachments/assets/a59ee5c0-6bcb-4f8a-bba4-97c53425cf0d" />

**Comment by isaackhlam** [comment]
Hi, today I try use the slider table, and it works. But seems the step is depended on the internal step counter? May I ask if it is possible to get my own step counter?

```python
run.define_metric("table", step_metric="epoch")

for epoch in range(5):
   # mimic step for batch data
    for i in range(10):
        run.log({"step_loss": i/10})

    table = wandb.Table(columns=['k1', 'k2'])
    table.add_data({'k1': 1, 'k2': 2})

    run.log({"table": table, "epoch": epoch})
```

And the step slider I got on the console is indexed with step 11, 22, ... rather than 1, 2, ... 
If I do `run.log({"table": table, "epoch": epoch}, step=epoch)` it complains about
> wandb: WARNING Tried to log to step 1 that is less than the current step 10. Steps must be monotonically increasing, so this data will be ignored. See https://wandb.me/define-metric to log data out of order.

Thanks for anyone could help.
